### ISSUE TITLE
redundant_closure_call report on const closure within non const test fn, which is not redundant

### ISSUE THREAD

**Issue by loynoir** [rust-lang/rust-clippy#14185]
### Summary

redundant_closure_call report on const closure within non const test fn, which is not redundant

### Lint Name

redundant_closure_call

### Reproducer

I tried this code:

```rs
#![feature(const_closures)]
```

```rs
    #[test]
    pub fn should_const_hello() {
        #[expect(clippy::redundant_closure_call)]
        let actual = (const || hello(42))();

        non_const_assert_eq(actual, 42);
```

I saw this happen:

```rs
#[expect(clippy::redundant_closure_call)]
```

I expected to see this happen:

```rs
// #[expect(clippy::redundant_closure_call)]
```

### Version

```text

```

**Comment by samueltardieu** [comment]
Thanks for your report. Could you please provide a minimal example that can be compiled which demonstrates what you think is a bug?

**Comment by loynoir** [comment]
```rs
#![expect(incomplete_features)]
#![feature(const_closures)]

pub const fn add(left: u64, right: u64) -> u64 {
    left + right
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_works() {
        #[expect(clippy::redundant_closure_call)]
        let result = (const || add(2, 2))();

        assert_eq!(result, 4);
    }
}
```

```rs
        let result = (const || add(2, 2))();
```

clippy think above code is redundant, and suggest

```rs
        let result = add(2, 2);
```

But these two are not same.

- First one restrict `add` is const fn

- Second one does not restrict `add` is const fn

**Comment by samueltardieu** [comment]
I'm not sure I understand what bugs you here:

- The suggested fix will give the same result given the current code base, but makes no guarantee that if the code around it is modified then errors will manifest the same way.
- There are other ways to ensure that `add` is a `const fn` other than defining then calling a closure, for example by defining
  ```rust
  let call_result = const || add(2, 2);
  …
  assert_eq!(call_result(), 4);
  ```

Similarly, you could say that rustc should not lint

```rust
const fn dummy() {
    let _ = add(0, 0);
}
```
as being unused and suggest its removal because you use it to ensure that `add()` is a `const fn` through this unused definition.

Or did I miss your point?

**Comment by loynoir** [comment]
Oh, new idea, there are block within rust, const block seems working.

```rs
let _ = const { const_add(42, 42) };
```

**Comment by samueltardieu** [comment]
> Oh, new idea, there are block within rust, const block seems working.
> 
> let _ = const { const_add(42, 42) };

Yes, but be careful, doing so calls the function at compile time and may have side effects: for example if the function panics, it will stop the compilation. So this is not equivalent to just checking that it is const. Hence my proposal with `call_result` above which doesn't make the call.
